### ISSUE TITLE
The result is not so good at my own dataset , especially small object, any suggestion?

### ISSUE THREAD

**Issue by hwh-hit** [dbolya/yolact#201]
hi,thanks for your great work. I have trained my own dataset with yolact. the bakbone are resnet101 and resnet50,iters=100000, my dataset have 6000 images.But the result is not so good for small object.Is there anything i can do to improve the result ? My image size is 1280x720, in config.py, i set max_size=550.should i resize the image first?
![yolact_](https://user-images.githubusercontent.com/34235323/68261460-2cc05100-007b-11ea-8a9e-57c9a7b95a06.png)

**Comment by hwh-hit** [comment]
By the way, when i turn off the data augmentation in resnet101, i get the result
![yolact__](https://user-images.githubusercontent.com/34235323/68263666-7b70e980-0081-11ea-802b-4c1d81d1a4f8.png)

**Comment by abhigoku10** [comment]
@hwh-hit @dbolya  i have similar results for lane detection for my custom dataset , @hwh-hit thanks for sharing the results . i get broken masks most of the time . @hwh-hit what dataset are you using for lane detection ??

**Comment by hwh-hit** [comment]
@abhigoku10 not any standard dataset， i made it by myself, labelme(json)->coco(json),  not only one label.  i have tried maskrcnn whose backbone is mobilenet with the same dataset , but the img size is 384x384, not so accurate ,but the mask is a complete one，so i can process the result with some image process algorithms. so i am here to ask some help for @dbolya to improve the performance.
![yolact___](https://user-images.githubusercontent.com/34235323/68267354-1fac5d80-008d-11ea-91dd-40f5e143f4ec.png)

**Comment by abhigoku10** [comment]
@hwh-hit  the above results is from mask rcnn output ??. with yolact i was also facing the same issue with the mask generation

**Comment by hwh-hit** [comment]
@abhigoku10 yes，the last image is the output of maskrcnn.

**Comment by abhigoku10** [comment]
@hwh-hit nice check out bbd100k dataset it has lane detection category dataset

**Comment by dbolya** [comment]
Hmm it looks like there is regular artifacting in YOLACT's masks. Can you send me one of your original images so I can test to see what specific part of YOLACT might be causing this artifacting? Also your trained weights would be helpful.

The procedure for predicting masks starts with resizing all images to 550x550, predicing the masks, and then resizing the 138x138 masks back to the original image size (in your case, 1280x720). The artifacts may be being caused in the scaling method for the last step, since if you look at some of the lines and imagine them squished horizontally, they form contiguous lines.

**Comment by hwh-hit** [comment]
@dbolya ok，i see, but i am so sorry to say that the trained weight is to big，185M,  i can‘t upload here, here is my email:991185721@qq.com, please send me a email, i will send images and trained weigthts to you

**Comment by hwh-hit** [comment]
thanks a lot ,i will have a try @abhigoku10

**Comment by dbolya** [comment]
@hwh-hit Okay sent

**Comment by hwh-hit** [comment]
@dbolya Okay sent, please let me know if you need more details or images. looking forward to your reply.

**Comment by dbolya** [comment]
@hwh-hit Okay received. I'll take a look at this tomorrow.

**Comment by cendelian** [comment]
> hi,thanks for your great work. I have trained my own dataset with yolact. the bakbone are resnet101 and resnet50,iters=100000, my dataset have 6000 images.But the result is not so good for small object.Is there anything i can do to improve the result ? My image size is 1280x720, in config.py, i set max_size=550.should i resize the image first?
> ![yolact_](https://user-images.githubusercontent.com/34235323/68261460-2cc05100-007b-11ea-8a9e-57c9a7b95a06.png)

hi could you tell me what's your GPU?
and how much your gpu memory is？
i meet some error for my pool gpu

**Comment by abhigoku10** [comment]
@hwh-hit @dbolya  i have observed the similar mask issue with HTC network architecture ,is this caused due to resizing of mask for the extended resoluiton

**Comment by hwh-hit** [comment]
@dbolya @abhigoku10 To find the reason, I show one of the 138 x 138 x num_mask pictures, then resize it to 1280x720, it's not exactly same as the output of yolact.
part of 138 x 138
![mask_y1](https://user-images.githubusercontent.com/34235323/68359563-420aad80-0157-11ea-89c8-97fb4891878d.png)
part of 1280 x 720
![mask_y2](https://user-images.githubusercontent.com/34235323/68359871-b8f47600-0158-11ea-9123-c40e56d7be33.png)

**Comment by abhigoku10** [comment]
@hwh-hit i think when we resize the mask its bad, we need to have mask rescaling logic upgraded

**Comment by dbolya** [comment]
So I see two options here:
1. Use a different scaling method as @abhigoku10 suggested. This is fairly simple to try. All you need to do is change this line: https://github.com/dbolya/yolact/blob/9730ab094e70461a8ee0f5f428b1821e29f3b3e8/layers/output_utils.py#L15 Instead of 'bilinear', try 'nearest'. Pytorch might complain about `align_corners=True`. If so just ctrl+f and remove all instances of that. I believe this is the scaling mode you used to resize that image yourself, so it might be more what you want.
2. Instead of using a square 550x550 image as input, use something with aspect ratio 16:9. There's no reason why YOLACT shouldn't work with a non-square image (so long as that image is constant size), but my config simply doesn't support it at the moment. I think the rest of my code does, however, so there should just be a couple of changes necessary. If you think you want to try that, I can implement it (but try (1) first).

**Comment by abhigoku10** [comment]
@dbolya @hwh-hit  i tried the first approach , but the results did not improve i used to get lot of boxes of mask , @dbolya is it possible to implement for  non square images training and testing . what will be draw back by having non -square images

**Comment by liudanyang** [comment]
> @abhigoku10 not any standard dataset， i made it by myself, labelme(json)->coco(json), not only one label. i have tried maskrcnn whose backbone is mobilenet with the same dataset , but the img size is 384x384, not so accurate ,but the mask is a complete one，so i can process the result with some image process algorithms. so i am here to ask some help for @dbolya to improve the performance.
> ![yolact___](https://user-images.githubusercontent.com/34235323/68267354-1fac5d80-008d-11ea-91dd-40f5e143f4ec.png)

Hi~Could you please share your labelme->coco script~Many thanks!

**Comment by dbolya** [comment]
@abhigoku10 I'll add it to the todo. I think it shouldn't be too bad so I'll probably be able to do it in the next couple of weeks.

The downsides would be stretched anchor boxes and maybe more pixels so its slower. You could fix both of those issues pretty easily though just by changing the config.

**Comment by abhigoku10** [comment]
@hwh-hit @dbolya  I have trained lane detection and i get output as shown below , should i re-generate the anchors to get different bounding box ?? Most of the masks for these lanes are continuous how to get a better masks ??
![image](https://user-images.githubusercontent.com/22094053/71799922-a1643b80-307c-11ea-8c50-791bc34b06cf.png)

**Comment by dbolya** [comment]
@abhigoku10 Honestly your best bet barring rescaling is to add extra upsample layers for bigger masks.

To do that, on this line:
https://github.com/dbolya/yolact/blob/2a11442dab4a5b566cf3ce36ef178e45e146b2d5/data/config.py#L691
change `[(None, -2, {}), (256, 3, {'padding': 1})]` to `[(None, -2, {}), (256, 3, {'padding': 1})] * 2` to double the mask size, `*3` to quadruple, etc. Note that this will take more vram so you might need a smaller batch size, though.

**Comment by abhigoku10** [comment]
@dbolya but during inference the performance will decress right if i add extra upsample layers for bigger masks

**Comment by dbolya** [comment]
@abhigoku10 Yeah it'll decrease. It might be not that much though, because you're only adding one extra layer on top of what would already be upsampled to the image resolution. So the actual performance hit might be negligible. I'd honestly worry more about the memory usage.

**Comment by hwh-hit** [comment]
Hi @dbolya , 
Is there any update on this issue? Looking forward to your reply.

**Comment by abhigoku10** [comment]
@hwh-hit  i did not make any changes in the architecture but i used yolact ++ i was getting acceptable results . did u do any modifications to the network ???

**Comment by hwh-hit** [comment]
Hi, @abhigoku10 , 
I was troubled with something else, so sorry, I didn't make any modifications. 
And could you post some result pictures here?

**Comment by abhigoku10** [comment]
@hwh-hit  please find the output below . but i have few issues with the mask during some scenes
![image](https://user-images.githubusercontent.com/22094053/75170353-fb65b100-574f-11ea-8b11-90193623894c.png)

**Comment by dbolya** [comment]
@hwh-hit I'll be adding fixed non-square image support soon. I'm just currently swamped with paper writing and other work.

**Comment by hwh-hit** [comment]
@abhigoku10 Thanks for your reply, the result seems better than the former one.  Are you using resnet as backbone?  And Is the dataset made by yourself or a public dataset?  The drivable area looks good.
@dbolya Ok, thanks for your great work again.

**Comment by abhigoku10** [comment]
@hwh-hit i am using resnet 50 backbone , its the public availabe dataset which was manually annotated

**Comment by qjziyou** [comment]
> @dbolya @abhigoku10 To find the reason, I show one of the 138 x 138 x num_mask pictures, then resize it to 1280x720, it's not exactly same as the output of yolact.
> part of 138 x 138
> ![mask_y1](https://user-images.githubusercontent.com/34235323/68359563-420aad80-0157-11ea-89c8-97fb4891878d.png)
> part of 1280 x 720
> ![mask_y2](https://user-images.githubusercontent.com/34235323/68359871-b8f47600-0158-11ea-9123-c40e56d7be33.png)

@dbolya  I have the same problem as @hwh-hit.  I use yolact++ on person segmentation task. when i resize the mask from 138X138 to 1920X1080, there are aliasing on the edge of resized mask.  however, i  test maskrcnn with the output  size 112X112 later, and resize the mask to  1920X1080, there is no aliasing. I use the same dataset(my own dataset) to train the two models. what's the different between these two nets. will the method non-square image solver this problem?

**Comment by dbolya** [comment]
@qjziyou That's interesting. Can you show me example outputs from both Mask R-CNN and YOLACT for the same image?
